TypeScript signatures for core-data selectors

adamziel opened the pull request:
## Description

This PR proposes a set of type signatures for selectors in core-data.

Because of the sheer volume of changes, I'm splitting this large PR into a series of smaller, atomic changes:

* https://github.com/WordPress/gutenberg/pull/38666
* https://github.com/WordPress/gutenberg/pull/39349
* https://github.com/WordPress/gutenberg/pull/40025
* https://github.com/WordPress/gutenberg/pull/40024
* https://github.com/WordPress/gutenberg/pull/40026
* https://github.com/WordPress/gutenberg/pull/40995
* https://github.com/WordPress/gutenberg/pull/41235
* https://github.com/WordPress/gutenberg/pull/41593
* https://github.com/WordPress/gutenberg/pull/41595

Overall, the challenges here can be illustrated using `getEntityRecord` as an example:

https://github.com/WordPress/gutenberg/blob/0b1df8882563ca8f68086955ccbd3ba3bea78110/packages/core-data/src/selectors.js#L127-L136

And the type signature proposed in this PR:

```ts

export type getEntityRecord = <
	R extends RecordOf< K, N >,
	C extends Context = DefaultContextOf< R >,
	K extends Kind = KindOf< R >,
	N extends Name = NameOf< R >,
	Q extends {
		/**
		 * The requested fields. If specified, the REST API will remove from the response
		 * any fields not on that list.
		 */
		_fields?: string[];
	} = {}
>(
	state: State,
	kind: K,
	name: N,
	key: KeyOf< R >,
	query?: Q & {
		context?: C;
	}
) =>
	| ( Q[ '_fields' ] extends string[]
			? Partial< RecordOf< K, N, C > >
			: RecordOf< K, N, C > )
	| null
	| undefined;
```

Entity record types are associated with string-based kind and name. Type inference is enabled through `CoreEntity`, a database-disguised-as-a-lookup-type that enables looking up (kind,name)->recordType and vice versa.

Typing the entire core-data is way beyond this PR. This proposal paves the way by adding the appropriate type declarations to `src/types/selectors.ts` without bringing them over to `src/selectors.js` yet.


### Test plan

Create a scratch file inside core-data and confirm the following data types check out:

```ts
const commentDefault = getEntityRecord( {}, 'root', 'comment', 15 );
// commentDefault is Comment<'edit'>

const commentView = getEntityRecord( {}, 'root', 'comment', 15, {
	context: 'view',
} );
// commentView is Comment<'view'>

const commentInvalidPK = getEntityRecord( {}, 'root', 'comment', '15' );
// commentInvalidPK shows a TypeScript error

const commentCustom = getEntityRecord<Comment, 'view'>({}, 'root', 'comment',15, { context: 'view' });
// commentCustom is Comment<'view'>
```

The `R` generic is the entity record type, and is the primary mean of configuring the calls to `getEntityRecord`. The `C`, `K`, and `N` generics would ideally be removed, but I don't see an easy way to do that.

cc @dmsnell @jsnajdr @sarayourfriend @sirreal

adamziel committed:
Propose type signatures for core-data selectors

adamziel committed:
Exploring different ways of accessing entity record types

dmsnell commented:
Whenever I'm working on complicated types like this I like to watch my own code and check for an abundance of special-casing and awkward conventions and if I see that, think about if I'm pushing too hard on approach vs. stepping back to check if there might be others.

We have two really obvious approaches with `getEntityRecord`:
	- auto-type the entire thing with no type annotations necessary on the part of the callers
	- rely on manually providing the type of the returned object

In this PR we're shooting for the first one, but obviously there's a ton of machinery in there to make it work, and some remaining questions about the ease of extending it. There are questions about the relationship between `kind` and `name` (or maybe those are obvious and it's just my ignorance making me wonder). The types can be a maze, and you found a way through the maze to get the desired output, and bravo for that 👏. Still, it's a winding path with lots of type helpers and wrappers.

The other approach is almost too simple to warrant discussion.

```ts
export const getEntityRecord = <T extends EntityRecord<any>>(…): T | null => { … }

const post = getEntityRecord<Post<'edit'>>('postType', 'post', 13, { context: 'edit' } );
```

Every question about extending this is already taken care of because it's up to you to supply the return type (as the caller).

I hope that we can get to full auto-typing and inference, but what I'm wondering is if there's a way to step there that starts with the simpler approach, specifically of having the `EntityRecord<any>` type parameter as the first parameter. That would let us move upwards to more inference; in the other direction though, it would be awkward to have to write this if we had to give up and step down to the manual approach:

```ts
const post = getEntityRecord<'postType', 'post', 13, 'edit'>('postType', 'post', 13, { context: 'edit' } );
```

Is this possible to try and look for a type signature for full inference that starts with `T extends EntityRecord<C>`? I don't know. Maybe if not there's at least `<C extends Content, T extends EntityRecord<C>>`

```ts
export const getEntityRecord = <C extends Context, T extends EntityRecord<C>>(
	state: State,
	kind: 
)
```

For example, I believe that this gets us the expected `kind` given an entity type.

```ts
type KindOf<T extends EntityRecord<any>> = keyof OmitNevers<{
	[Kind in keyof EntityRecordLookup['root']]:
		EntityRecordLookup['root'][Kind] extends T
			? T
			: never;
}>;
```
![Screen Shot 2022-02-23 at 12 50 44 PM](https://user-images.githubusercontent.com/5431237/155397015-75ee641e-20c1-46d9-a61b-de38c3d4ab17.png)

The great thing about extensibility is that if we can get to the point where this is parameterized by one required type parameter, we should be able to provide automatic inference for all the core types without imposing any difficult steps on people to extend it. Maybe they don't want to go overwrite ambient namespaces types; they can supply their own type parameter and call it a day. We can potentially do this through conditional types checking if the given EntityRecord is known or not.

```ts
export const getEntityRecord = <T extends EntityRecord<any> | CustomEntityRecord>(…): T | null { … }
```
![Screen Shot 2022-02-23 at 12 56 54 PM](https://user-images.githubusercontent.com/5431237/155397907-d642572b-5249-439b-a0f5-d9f4f67d926d.png)

https://user-images.githubusercontent.com/5431237/155398611-44b5e15a-7beb-49e1-991a-2f9a54b10fab.mov

Lots of doodling, but I'm just trying to see how far we can get with how little in the terms of type parameters and annotation.

adamziel committed:
Simplify the types

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -216,144 +216,194 @@ export const defaultEntities = [
+			recordType: WpTemplatePart< C >;
+			keyType: WpTemplatePart< C >[ 'id' ];
+			defaultContext: 'edit';
+	  };
```
quick note: it's generally convenient to define these separately and then join them together by name. same effect on the types but the names are handy as separate things.

```ts
// entity-type-definitions.ts
export interface Settings { … }

…

export EntityDefinition =
	| Settings
	| Type
	| Attachment
	| …
```

I know it's early and you don't want to focus on the specifics of the type, just wanted to point this out

adamziel committed:
More succinct EntityTypeUnion declaratio

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -216,144 +216,194 @@ export const defaultEntities = [
+export type Kind = EntityTypeUnion[ 'kind' ];
+export type Name< K extends Kind > = ( EntityTypeUnion & {
+	kind: K;
+} )[ 'name' ];
```
does this work? I believe we can also use [`Extract`](https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union) to "find' the right value in there

```ts
export type Name< K extends Kind > = Extract< EntityTypeUnion, { kind: K } >['name'];
```

[playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMJzMg3gKGf5Aa1ABMAuZAchggBtQJKBuPAgZzgE82LKBbCAHsA7sxwBfHDlCRYiFABFBAc2yt8xEOSoI4IBmILIO3XgCM4UQmMk4wnAA4oAgvr5xayALxoMyAD7ISsosdo4oAMpcbAA8ANKkyBAAHpBabMiuwO60ANqUmiSUALoAfN7IAKKpUIhgMZluHgA02ESkFAlayOLIpfkmbCWh9k7IALJCwgJQGT5R3DHUdAalLAD06wQAegD8QA)

adamziel replied:
Extract worked great, thank you!

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -136,33 +188,125 @@ export const defaultEntities = [
+	defaultContext: C;
+};
+
+export type EntityTypeUnion< C extends Context = any > =
```
nitpick: we don't need `Union` in the name. Could be something like `EntityTypeDefinition` or `DeclaredEntity` or `EntityDeclaration`

adamziel committed:
Yes! Correct type inference in getEntityRecord

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -13,121 +13,173 @@ import { __ } from '@wordpress/i18n';
+	WidgetType,
+	WpTemplate,
+	WpTemplatePart,
+} from './types';
```
fun fact! we can import these altogether and reference them as if they were properties of the module.

```ts
import type * as EntityTypes from './types';

…
| DeclaredEntity< 'root', 'site', EntityTypes.Settings< C > >
```

this can free up the local module's namespace for those same entity kinds

adamziel replied:
Good idea!

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -13,121 +13,173 @@ import { __ } from '@wordpress/i18n';
+	context?: C;
+};
+
+export interface EntityDefinition extends Object {
```
is there a reason to extend `Object`?

adamziel replied:
Not really, that was just an artifact from some previous iteration – I removed it

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -136,33 +188,126 @@ export const defaultEntities = [
+	| APIEntity< 'postType', 'post', Post< C > >
+	| APIEntity< 'postType', 'page', Page< C > >
+	| APIEntity< 'postType', 'wp_template', WpTemplate< C > >
+	| APIEntity< 'postType', 'wp_template_part', WpTemplatePart< C > >;
```
this is great. I love pushing the computation into the type system and when it matches the way we think about the actual code.

adamziel committed:
Simplify Name handling

adamziel committed:
remove extends Object

adamziel commented:
My brain is fried so just a quick note – I think the full inference works (although it's based on kind and type)! 

![CleanShot 2022-02-23 at 22 27 15](https://user-images.githubusercontent.com/205419/155411528-f7fa62bd-3484-45c6-8d88-0bad3f670412.png)
![CleanShot 2022-02-23 at 22 27 24](https://user-images.githubusercontent.com/205419/155411534-2d81e88f-58e5-4f60-9890-f02ccb72835d.png)

And the types aren't even that complicated:

```ts
export type EntityType< C extends Context = any > =
	| DeclaredEntity< 'root', 'site', Settings< C > >
	| DeclaredEntity< 'root', 'postType', Type< C > >
	| DeclaredEntity< 'root', 'media', Attachment< C > >
	| DeclaredEntity< 'root', 'taxonomy', Taxonomy< C > >
	| DeclaredEntity< 'root', 'sidebar', Sidebar< C > >
	| DeclaredEntity< 'root', 'widget', Widget< C > >
	| DeclaredEntity< 'root', 'widgetType', WidgetType< C > >
	| DeclaredEntity< 'root', 'user', User< C > >
	| DeclaredEntity< 'root', 'comment', Comment< C > >
	| DeclaredEntity< 'root', 'menu', NavMenu< C > >
	| DeclaredEntity< 'root', 'menuItem', NavMenuItem< C > >
	| DeclaredEntity< 'root', 'menuLocation', MenuLocation< C > >
	| DeclaredEntity< 'root', 'navigationArea', NavigationArea< C > >
	| DeclaredEntity< 'root', 'theme', Theme< C > >
	| DeclaredEntity< 'root', 'plugin', Plugin< C > >
	| APIEntity< 'postType', 'post', Post< C > >
	| APIEntity< 'postType', 'page', Page< C > >
	| APIEntity< 'postType', 'wp_template', WpTemplate< C > >
	| APIEntity< 'postType', 'wp_template_part', WpTemplatePart< C > >;

export type Kind = EntityType[ 'kind' ];
export type Name = EntityType[ 'name' ];

export type EntityRecordType<
	K extends Kind,
	N extends Name,
	C extends Context = any
> = Extract< EntityType< C >, { kind: K; name: N } >[ 'recordType' ];

export type Key< K extends Kind, N extends Name > = Extract<
	EntityType,
	{ kind: K; name: N }
>[ 'keyType' ];

export type DefaultContext< K extends Kind, N extends Name > = Extract<
	EntityType,
	{ kind: K; name: N }
>[ 'defaultContext' ];
```

My brain is fried so that's all I'm going to write right now, but my next comment will address the specific points you brought up @dmsnell 

--

edit:

Here's the full EntityType-based inference:

```ts
type Context = 'edit' | 'update' | 'view' | 'embed';

interface Post < C extends Context > { brand: 'post' }
interface Comment < C extends Context > { author: 2 }
interface Template < C extends Context > { distinct: 84 }
type EntityRecord < C extends Context > = Post<C> | Comment<C> | Template<C>
type State = any;

type EntityType< C extends Context > =
	| { kind : 'root', name: 'post', recordType: Post<C> }
	| { kind : 'root', name: 'comment', recordType: Comment<C> }
	| { kind : 'postType', name: 'template', recordType: Template<C> }

export type RecordOf< Kind, Name, C extends Context > = Extract<
	EntityType<C>,
	{ kind: Kind, name: Name }
	>['recordType']
export type KindOf< R extends EntityRecord<any> > = Extract<
	EntityType<any>,
	{ recordType: R }
	>[ 'kind' ];
export type NameOf< R extends EntityRecord<any> > = Extract<
	EntityType<any>,
	{ recordType: R }
	>[ 'name' ];

export function getEntityRecord<
	R extends RecordOf< Kind, Name, C >,
	C extends Context='edit',
	Kind extends EntityType<C>['kind']=KindOf<R>,
	Name extends EntityType<C>['name']=NameOf<R>
	>(
	state: State,
	kind: Kind,
	name: Name,
	query?: {
		context: C
	}
) : R { return {} as any; }

const post1 = getEntityRecord({}, 'root', 'post')
const post2 = getEntityRecord({}, 'root', 'post', { context: 'view' })
const post3 = getEntityRecord<Post<'edit'>>({}, 'root', 'post')
const post4 = getEntityRecord<Post<'view'>, 'view'>({}, 'root', 'post', { context: 'view' })

```

[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gO2BAHsKBeKByCATAS2GygB8cBXMPAQ2RPOwDcCIB3BnCAWwCN9sAbgBQwgkggAnAGY0AxtAAKcAM7oAPLCipkCPCtiJkaKAD4oAbyi9JNPQC4cYVcSgBfMRJnzo8btwgkKE0YbTRA-UMJE3MrGgpgAAs4SUcAJndPZG8FKAAVHjAAGzpoELDdSPho9FioQjVxOWBHAA4AFkzQSCgAUSQiEAAlCDkUvGCtHQiDauNazChlNXUYc3I-AKRV9fzCkuQd4W7oAGVgUsW7EBFj8Gh+4EG8+-LpvVmjHTNF4Sh-siWKAAa3EE0c2EkcDgxAANFAEDQAhDnGpsPDJKNxi9II5lsAdpkAYCrKC9FAIVCYeiEUiIBCxv5AnCoJixpI8Dj6YYmds1kSAeRSWCKU4XFyaYjkThkNxiqUaWzsfdHAU5QcIISPMJUM5JOgTlARuy8AB5aSaADSYPhADk6fDQu8ql8YotemhbM11H8AY9nq81rDff9hQ4oNa9PCpdz7QFMqYANqQrEciUAXR1KD1BvuEbB5s0QwqMz6A1AxvG6mu5nMWA9wC9BJDZaeoC51YQIFMweJViVaZVRoTiZwZLwJHTIl1KVzPTjEELw+dBn9FdTeE73Z+9c98mbxLXIA7Nd7AP7G65jmLHiTOBjk9uM-1UGkFAQzQIiCgAHMIMAj0rDkfWJYsVyNDcl0jPA7QdLQexbJ1wg+KJ5gwXBCDhFtoJLFCjw7NZk3HbB0wwaDCyGBDiQXXDInwwMk2wB9SIXCjTGEUwAAoWzUUpHHOUoz3+cdHGgoTaWlBdxIARwoKQQAAfkcCwW2JMYakcGAWw8ABKUViwvYAKEkBBLDcKAaAMa5BEyYR1LUKBUWAABGRY-wA8thg3TiLDceFIWhFlsCc7AdLsxAHKcjIsHcwDvN8-yqSCkL4SsdT5ghFh2BINwwvs9AnIAZjc-84pNdR8XUDCiGwUwuISnAkppYKXFC8KEEilxOhi0rPKAzdKuYVgOB7HAspGny-MawLmpSoF0p0TLhpysKgA)

sarayourfriend commented on packages/core-data/src/wordpress-data.d.ts:
```diff
@@ -0,0 +1,3 @@
+declare module '@wordpress/data' {
+	export function createRegistrySelector< T extends (selector) => F, F extends Function >( fn: T, ...any ) : F;
+}
```
Nice stop-gap :+1:

sarayourfriend commented on packages/core-data/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+	"noUnusedParameters": false
+  },
+  "references": [
+  ],
```
```suggestion
  "references": [
    { "path": "../i18n" },
    { "path": "../api-fetch" },
  ],
```

sarayourfriend commented on packages/core-data/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+  "references": [
+  ],
+  "include": [
+	"src/types/**/*",
+	"src/*.ts"
```
```suggestion
	"src/**/*.ts"
```

adamziel replied:
Thanks! I removed the tsconfig.json for now, but this will come handy later on

adamziel committed:
Entity type-based inference

adamziel committed:
Isolate the getEntityRecord selector

adamziel commented:
I reduced this PR to focus just on the `getEntityRecord` part of it.

> The other approach is almost too simple to warrant discussion.
> 
> ```ts
> export const getEntityRecord = <T extends EntityRecord<any>>(…): T | null => { … }
> 
> const post = getEntityRecord<Post<'edit'>>('postType', 'post', 13, { context: 'edit' } );
> ```

Great point! This would "just work" and solve a ton of problems here.

The full inference is just so tempting, though 😆 I reduced this PR to something much simpler than it used to be, combining the first version with your explorations, which enabled the following signature:

```ts
export const getEntityRecord = function <
  R extends EntityRecordType< K, N, C >,
  K extends Kind = KindOf< R >,
  N extends Name = NameOf< R >,
  C extends Context = DefaultContextOf< K, N >
>(
  state: State,
  kind: K,
  name: N,
  key: PrimaryKey< R >,
  query?: EntityQuery< C >
): R | null | undefined 
```

Unfortunately, I couldn't the inference to work right without the K and N generic parameters. This how it is used:

```ts
const commentDefault = getEntityRecord( {}, 'root', 'comment', 15 );
// commentDefault is Comment<'edit'>

const commentView = getEntityRecord( {}, 'root', 'comment', 15, {
	context: 'view',
} );
// commentView is Comment<'view'>

const commentInvalidPK = getEntityRecord( {}, 'root', 'comment', '15' );
// commentInvalidPK shows a TypeScript error

const commentCustom = getEntityRecord< Comment< 'edit' > >( {}, 'root', 'comment', 15 );
// commentCustom is Comment<'edit'>

const commentCustomContext = getEntityRecord<
	Comment< 'view' >,
	'root',
	'comment',
	'view'
>( {}, 'root', 'comment', 15, {
	context: 'view',
} );
// commentCustom is Comment<'view'>
```

It requires the K and N parameters when you want to pass a custom Record type in a non-edit context.

I'm not sure the `R` generic type can be simplified to `<T extends EntityRecord<any>>` – we need something to somehow tie it to the kind and name function arguments and I couldn't get it to work like this:

```ts
// No bueno in either case
export const getEntityRecord = function <
	R extends EntityRecord< C >,
	C extends Context = 'edit'
>(
	state: State,
	kind: KindOf< R >,
	name: NameOf< R >,
	key: PrimaryKey< R >,
	query?: EntityQuery< C >
): R

export const getEntityRecord = function <
	R extends EntityRecord< C >,
	C extends Context = 'edit'
>(
	state: State,
	kind: KindOf< R >,
	name: NameOf< R >,
	key: PrimaryKey< R >,
	query?: EntityQuery< C >
): EntityRecordType< KindOf< R >, NameOf< R >, C > 
```

cc @dmsnell @sarayourfriend

sarayourfriend commented on packages/core-data/src/ts-selectors.ts:
```diff
@@ -0,0 +1,137 @@
+ * yet received, undefined if the value entity is known to not exist, or the
+ * entity object if it exists and is received.
+ *
+ * @param {Object}  state State tree
+ * @param {string}  kind  Entity kind.
+ * @param {string}  name  Entity name.
+ * @param {number}  key   Record's key
+ * @param {?Object} query Optional query.
+ *
+ * @return {Object?} Record.
```
```suggestion
 * @param state State tree
 * @param kind  Entity kind.
 * @param name  Entity name.
 * @param key   Record's key
 * @param query Optional query.
 *
 * @return Record.
```

sarayourfriend commented:
> The full inference is just so tempting, though laughing I reduced this PR to something much simpler than it used to be, combining the first version with your explorations, which enabled the following signature:

Keeping in mind that most WordPress contributors will probably be consuming this from regular JS and not have the ability to pass type parameters to functions, full inference would be a good goal to reach; we don't even have full TS support in the create-block generator yet.

adamziel commented:
I boiled the generics signature down to:

```ts
<
    R extends RecordOf< Kind, Name, C >,
    C extends Context=RContextOf<R>,
    Kind extends EntityType<C>['kind']=KindOf<R>,
    Name extends EntityType<C>['name']=NameOf<R>,
>
```

Which is close to perfection – the record comes first, then the optional context, and then optional Kind and Type which can almost always be omitted. Here's the usage:

```ts

const post1 = getEntityRecord({}, 'root', 'post')
// post1 is Post<'edit'>

const post2 = getEntityRecord({}, 'root', 'post', { context: 'view' })
// post2 is Post<'view'>

const post3 = getEntityRecord<Post<'edit'>>({}, 'root', 'post')
// post3 is Post<'edit'>

const post4 = getEntityRecord<Post<'view'>, 'view'>({}, 'root', 'post', { context: 'view' })
// post4 is Post<'view'>
```

[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gO2BAHsKBeKByCATAS2GygB8cBXMPAQ2RPOwDcCIB3BnCAWwCN9sAbgBQwgkggAnAGY0AxtAAKcAM7oAPLCipkCPCtiJkaKAD4oAbyi9JNPQC4cYVcSgBfMRJnzo8btwgkKE0YbTRA-UMJE3MrGgpgAAs4SUcAJndPZG8FKAAVHjAAGzpoELDdSPho9FioQjVxOWBHAA4AFkzQSCgAUSQiEAAlCDkUvGCtHQiDauNazChlNXUYc3I-AKRV9fzCkuQd4W7oAGVgUsW7EBFj8Gh+4EG8+-LpvVmjHTNF4Sh-siWKAAa3EE0c2EkcDgxAANFAEDQAhDnGpsPDJKNxi9II5lsAdvC8BBZBQisA5joIfgiCQPADAVZQXooBCoTD0QikRAIWN-IE4VBMWNJHgcTzDPztmsiST4uTKWgISx2HS-gDyEywaynC5xZzEcicMhuMVSpzhdj7o4CqaDhBCfU5WSKV8lTgVRxMsJUM5JOgTlARiK8AB5aSaADSYPhADlufDQu8qm6FlhemhbM11Or-o9nq8Zbmgcy8I5o3p4YaJfGAplTABtSFY0X6gC6PpQfoD9ygFbDEaDFRmfQGoGD43U13M5nTmfkBOL+dA4qnCBAplhxasltb1qHHkbOFLJDbIl9KR7PVrEHDmiGw4+o6e45beDXG5+c+AWcXDOXICrtOW4Mjub7io4D6Hg2ODVqetyBkMirAHeD7JgYAETqKH6mLOfTztmS5joBrzAduQrgfuUHCEe2DEqSCqpvBogXv6UDSBQCDNAQiBQAA5hAwCYW+OYMmh4RPlhA5RjGUA3omZggQCSYSSmNQYEhqaoZuxb9o+kQAauaxNiebYYP22lKf8N76RhxFGY22BwWZN6WTRAAUxZqKUjjnKUVkgmC5YxsW1aOPJxYAI4UFIIAAPyOBYxYMmMNSODAxYeAAlDqD5gcAFCSAglhuFANAGNcgjeqlahQKiwAAIyLAJQnEVJ7kWG48KQtCgrYPV2BZcIAD0w11S4TUEAY+LqLghDEKYog1eg9UZFgLXCSGHVdTg7J9QN8JWKl8zKqwXpuENo3jWoGRTUsLizZ62CLcIy3XcAADMzWCZtk4zXNtK4dt3V7Zy-UuINI1jfVX13f9NILUtiC1fVnTrT9bUif9T2bh6Z3PcDu29WDB1AsdVJ46q7iXdDLidHDD3MPjphAA)

One downside I see is that the R parameter is not a `EntityRecord< C >` but `RecordOf< Kind, Name, C >`. They coincide, but `EntityRecord` is a type union while `RecordOf` is sourced from a different union:

```ts
type EntityRecord < C extends Context > = Post<C> | Comment<C> | Template<C>

type EntityType< C extends Context > = 
    | { kind : 'root', name: 'post', recordType: Post<C>, defaultContext: 'edit' }
    | { kind : 'root', name: 'comment', recordType: Comment<C>, defaultContext: 'view' }
    | { kind : 'postType', name: 'template', recordType: Template<C>, defaultContext: 'view' }

// RecordOf depends on EntityType
```

And this not extensible yet. We can fix that, though, by storing the entity types in an interface as below:

```ts
// core-data code
type CoreEntityType< C extends Context > = 
    | { kind : 'root', name: 'post', recordType: Post<C>, defaultContext: 'edit' }
    | { kind : 'root', name: 'comment', recordType: Comment<C>, defaultContext: 'view' }
    | { kind : 'postType', name: 'template', recordType: Template<C>, defaultContext: 'view' }

export interface EntityTypeWrapper< C extends Context > {
    core: CoreEntityType< C >
}

// Same as below, but now we read all the values stored in EntityTypeWrapper
type EntityType< C extends Context > = EntityTypeWrapper<C> [ keyof EntityTypeWrapper<C> ]

// This is no longer a hardcoded union type, but depends on the configuration
type EntityRecord < C extends Context > = EntityType<C>['recordType']

// Plugin code
interface Order { customerId: number; }

type myPluginEntityType< C extends Context > = 
    | { kind : 'myPlugin', name: 'order', recordType: Order, defaultContext: 'view' }

export interface EntityTypeWrapper< C extends Context > {
    myPlugin: myPluginEntityType< C >
}

const order = getEntityRecord({}, 'myPlugin', 'order')
// order is Order
```

In this case, `RecordOf< Kind, Name, C >` extends `EntityRecord< C >` which is what we're after, I believe.

One thing this setup doesn't allow is completely arbitrary return types as in `getEntityRecord<NotEntityRecord<'edit'>({}, 'myPlugin', 'order')` but I think that's fine.

dmsnell commented on packages/core-data/src/ts-selectors.ts:
```diff
@@ -0,0 +1,137 @@
+import type { Context } from './types';
+
+// Placeholder State type for now
+type State = any;
```
👍 it's nice seeing type aliases like this because they still appear in the annotations and documentation

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
-	},
-];
+	} as const ),
+] as const;
```
do we need the runtime function to wrap the types?
did you try this?

```ts
export const defaultEntities: EntityDefinition[] = [
	{ … },
	…
];
```

or even that, but with `] as const;` at the end instead of `];`

adamziel replied:
The type annotation didn't work, but I found a different workaround [on StackOverflow](https://stackoverflow.com/questions/57069802/as-const-is-ignored-when-there-is-a-type-definition). Tl;dr, it is possible to wrap the entire config in a function that guarantees type safety and extracts the type definition:

```ts
export const defaultEntities = entitiesConfig( [
	{
		label: __( 'Base' ),
		kind: 'root',
		name: '__unstableBase',
		baseURL: '/',
	},
	// ...
] );
// typeof defaultEntities is a union of specific literals
```

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
+	: 'view';
+
+type EntityKeyName<
+	K extends string,
```
for all these places with `string` I think we can use `Element['kind']` and `Element['name']` and this should give us auto-complete on the types, should we even need that. it should at least warn if we try to refer to one that doesn't exist and do so at compile-time

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
+	recordType: RecordType;
+	keyType: KeyName extends keyof RecordType ? RecordType[ KeyName ] : never;
+	defaultContext: C;
+};
```
what's the difference between a declared entity and an API entity and when and why should I choose one over the other?

adamziel replied:
Some configuration data is hardcoded in `entities.js`: https://github.com/WordPress/gutenberg/blob/a8ea1078c42a3312ee5402c2316086b50297a671/packages/core-data/src/entities.js#L21-L30

Other is loaded by requesting the API:
https://github.com/WordPress/gutenberg/blob/a8ea1078c42a3312ee5402c2316086b50297a671/packages/core-data/src/entities.js#L207-L215

With that in mind:
 
* DeclaredEntity is a name I came up with for the ones with a hardcoded configuration that we can directly reuse.
* APIEntity is a name for the entities with no TS-reusable configuration where we must explicitly specify the `defaultContext` and `key` – right now it always defaults to "edit" and "id".

I could clarify that via a docstring, but I wonder if you have an idea for a more intuitive naming scheme.

dmsnell replied:
is the distinction important? Do we need it at all?

adamziel replied:
I elaborated on that in https://github.com/WordPress/gutenberg/pull/39025#issuecomment-1054327402

dmsnell commented on packages/core-data/src/entities.ts:
```diff
@@ -12,157 +12,296 @@ import { __ } from '@wordpress/i18n';
+	| DeclaredEntity< 'root', 'plugin', EntityTypes.Plugin< C > >
+	| APIEntity< 'postType', 'post', EntityTypes.Post< C > >
+	| APIEntity< 'postType', 'page', EntityTypes.Page< C > >
+	| APIEntity< 'postType', 'wp_template', EntityTypes.WpTemplate< C > >
```
are these extra kind and name fields inherent to the entity records?
if so, should we consider adding them in the entity record types directly so we don't have to duplicate the list here? or are these values specific to `getEntityRecord` and not particularly a part of the entity record itself?

adamziel replied:
The API knows nothing about the kind and name fields, this is just a way for core-data to index them in the Redux state.

dmsnell commented:
because I have a bunch of changes in my local working copy I'm not going to jump into this now but will come back to it in a bit. I like where it's going; I want to see if I can figure out a way to get rid of the `Kind` and `Name` parameters entirely and I am 20% confident we can.

regardless this is going to be great and I'm glad we're pushing hard on some complicated internal types because of the way it will let plugin authors and core contributors outside of this package get all the type benefits with little to no annotation or work on their part.

sarayourfriend commented:
```
export interface EntityTypeWrapper< C extends Context > {
    core: CoreEntityType< C >
}

// Same as below, but now we read all the values stored in EntityTypeWrapper
type EntityType< C extends Context > = EntityTypeWrapper<C> [ keyof EntityTypeWrapper<C> ]
```

This is super clever, I would not have considered this approach at all. Definitely putting that one in my back pocket.

adamziel commented:
This is where we stand right now:

```ts
export const getEntityRecord = createSelector(
	function <
		R extends RecordOf< K, N >,
		C extends Context = DefaultContextOf< R >,
		K extends Kind = KindOf< R >,
		N extends Name = NameOf< R >
	>(
		state: State,
		kind: K,
		name: N,
		key: KeyOf< R >,
		query?: EntityQuery< C >
	): RecordOf< K, N, C > | null | undefined {
	// ...
);

const commentDefault = getEntityRecord( {}, 'root', 'comment', 15 );
// commentDefault is Comment<'edit'>

const commentView = getEntityRecord( {}, 'root', 'comment', 15, {
	context: 'view',
} );
// commentView is Comment<'view'>

const commentInvalidPK = getEntityRecord( {}, 'root', 'comment', '15' );
// commentInvalidPK shows a TypeScript error

const commentCustom = getEntityRecord<Comment<'view'>, 'view'>({},'root','comment',15,{ context: 'view' });
// commentCustom is Comment<'view'>
```

Any big-picture improvements you see? If we're in the right ballpark, I'll start cleaning this PR up.

dmsnell commented:
Still having a pretty hard time following the types around here and figuring out the purpose of each of them and how exactly to think through making changes or maintaining them.

I wonder if there's not a different approach that doesn't require so many layers of type wrapping. For instance, when trying to understand `DefaultContextOf` inside of the type for `getEntityRecord` I have to unwrap the following layers:

 - `DefaultContextOf`: okay this just finds a lookup entity record type in a metadata type and returns the `defaultContext`. Where's that stored?
 - `Entity`: seems to make sense, I assume it's `EntityRecord`, but it's not. It's `PerPackageEntity`, huh, okay, more metadata databasing.
 - `PerPackageEntity` is a think with `core` that I guess is for extensibility
 - `CoreEntity`: we should be getting here, but oh no, there are two types of types here and they don't stem from the same base type. I guess I'll try and ignore what `APIEntity` might be
 - `DeclaredEntity`: This has some type magic to lookup an entity config in an actual value `defaultEntities`. I _thought_ we were already past this lookup but it's happening again now on the value side of the equation. In any case it defines a `defaultContext` and sets the default value to `view`, which I thought should have been `edit`.

Maybe the type that `CoreEntity` plays could be a nested object instead of a type union?

If we're making an `EntityConfig` type anyway, can we not use that on `defaultEntities` and get what we want without the type wrapper?

Going to continue considering this. I'm a bit nervous about it the way it is at the moment, knowing we'll have to support it.

As always, thanks for your persistence and hard work on this!

adamziel commented:
> I wonder if there's not a different approach that doesn't require so many layers of type wrapping. For instance, when trying to understand DefaultContextOf inside of the type for getEntityRecord I have to unwrap the following layers:

@dmsnell You've definitely got a point here. I updated this PR to minimize the indirection as much as I could. I also made a distinction between entity _records_ and entity _types_ and added some nudges in the docstrings. I am not sure if this can be simplified much further besides shuffling things around.

> If we're making an EntityConfig type anyway, can we not use that on defaultEntities and get what we want without the type wrapper?

Using `defaultContext` as an example, we have three data sources to consult:

* `defaultEntities` declared in entities.js
* The knowledge we have about the post, page, wp_template, and wp_template_part entities that are not a part of `defaultEntities`
* Any additional information from the extenders

To make things more complex, `defaultEntities` is a value so the kind and type is not tied to the respective record type. For example, the following entry is a value and does not refer the `MenuLocation` record type:

```js
{
	name: 'menuLocation',
	kind: 'root',
	baseURL: '/wp/v2/menu-locations',
	baseURLParams: { context: 'edit' },
	plural: 'menuLocations',
	label: __( 'Menu Location' ),
	key: 'name',
}
```

In terms of databases, we are compiling the following data:

```sql
SELECT
	kind,
	name,
	recordType,
	DEFAULT(key, 'id') as key,
	DEFAULT(baseURLParams['context'], 'view') as defaultContext
FROM defaultEntities_declared_in_entities_js d
INNER JOIN record_types r ON d.kind = r.kind AND d.name = r.name
UNION
SELECT VALUES
	ROW ( 'postType', 'post', 'id', 'Post', 'edit' ),
	ROW ( 'postType', 'page', 'id', 'Page', 'edit' ),
	ROW ( 'postType', 'wp_template', 'WpTemplate', 'id', 'edit' ),
	ROW ( 'postType', 'wp_template_part', 'WpTemplatePart', 'id', 'edit' ),
AS what_we_intrinsically_know_about_these_unconfigured_entites
UNION
SELECT
	kind,
	name,
	recordType,
	key,
	defaultContext
FROM additional_entities_provided_by_extenders
```

This is quite a complex query and the underlying TS implementation reflects that complexity.

> Maybe the type that CoreEntity plays could be a nested object instead of a type union?

Do you mean something like below?

```ts
type CoreEntity = {
	root: {
		comment: Comment<C>,
		// ...
	}
}
```

That was my first instinct, too. I tried some variations of that, but it generated some additional hoops to jump through and the code ended up being more complex, not less.

I've also tried a few things like redeclaring the `key` and `defaultContext` in type definitions instead of sourcing them directly from entities.js, but then editing the configuration caused an unintuitive type errors along the lines of

> that context you just changed to `view` must be of type `edit`

dmsnell commented on packages/core-data/src/selectors-ts.ts:
```diff
@@ -0,0 +1,140 @@
+		query?: Q
+	):
+		| ( Q[ '_fields' ] extends string[]
+				? Partial< RecordOf< K, N, C > >
```
this is a great addition!

seeing `Partial` made me think we could go a step further and (for knowable cases) return an object that's not `Partial` but rather one that only has the requested fields.

surprisingly, I found a way to get TypeScript to infer that given the `_fields` value but I had the wrong type for `_fields` in my mind. I thought it was a comma-separated list of field names. I assumed it would be easier to do this given a list of field names, but I was wrong.

if people declare `_fields` with `as const` then we can do that. if not then I've only found a way to infer `string[]` as that field's type, meaning no information about the key names sadly.

adamziel replied:
@dmsnell I played with that a bit more and got some success with the following types:

```ts
type ExtractKeys<
	Source,
	Options extends { _fields?: ( keyof Source )[] }
> = Options extends {
	_fields: infer KeysArray;
}
	? KeysArray extends Array< infer KeysUnion >
		? KeysUnion
		: never
	: never;

type PickMaybeFn = < Source, Options extends { _fields?: ( keyof Source )[] } >(
	obj: Source,
	options?: Options
) => ExtractKeys< Source, Options > extends never
	? Source
	: Pick< Source, ExtractKeys< Source, Options > >;

const pick: PickMaybeFn = ( obj, options? ) => { /* ... */ };

const obj = { a: 1, b: 2, c: 3 };

// Return the verbatim object without the Pick wrapper
const dontPick: typeof obj = pick( obj );
const dontPick2: typeof obj = pick( { a: 1, b: 2, c: 3 }, { _fields: [] } );

// Return the Picked version of an object
const pickA: Pick< typeof obj, 'a' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a' ] }
);
const pickAB: Pick< typeof obj, 'a' | 'b' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a', 'b' ] }
);

// TS Error: 'c' is missing in type
const doNotPickAbc: Pick< typeof obj, 'a' | 'b' | 'c' > = pick(
	{ a: 1, b: 2, c: 3 },
	{ _fields: [ 'a', 'b' ] }
);
```

[Playground link](https://www.typescriptlang.org/play?ssl=28&ssc=1&pln=29&pc=1#code/C4TwDgpgBAogHsATgQwMbANIRAZwDwBQAkAMoD2AroqhADTEDyYwAlmQHY5QQITsAmXAN5QA+gDMWEADaCA-AC4oACigBrbGXFRyVGlACUAbQC6UAL4EAfFAC8UJqw5cewPoKhDiEqbJxKWdnEIRCgsXABBRBQQAG4CSyI5MOwcKJjuXgEudOQQPChA4NDwnABVdjZ2KCtiJJTcCqq6pXYIADcQ4laOkPiCUEgoAAUWVDUAWTyAIwgAMWr7At1qOgdmKpcsjxEfGXklVQ0QLR1KVcNTCxrlYjJpgCslFZp6IjIN50V1p04CAzsNngSDQmFSy3Orx+mxqmTc2SgbU6iGIyReEG6IzGaghejWwJQ6FKuNWtGhzlhVn6qGcwCgYGxSlG4ymIFmCzsKig9weZI+vxwyQBthsXiILG0qgAhNzPpw5AA6PZ+RXSPgAc2AAAtDJ46ogIMAqNUefEiIkaZw6QzxhB+EpkOwQJyhOYzeIyKFVJacHTjtztPzNkrJPsuACxUQbRp+EZ1NgoGZ7Dy4-6TGbEgajYhqtG7fE3QQCD66TyXVBkEoAIxk6ZKABMZNQSgAzBZ+gB6DtQABKhuNUG10GR02QrAAttzHhB0FAAO4sbWUOlDrHjecoMCQFElqD8DjAZlqJSDCCnMv2aOqMsGeK7-fsQ-Y+sn8BnwOPTlXzwV6u1htNq2FhkrsoZ+EoVzmIYnbdn22bVKuR52lAyI4FUAYVia07oMWtL0tiERMtiBSnuejxkgA5MgFGwpe2K3EQIiVlANZQHWUCNlAzZQG25hvKBviCBBUBUTRZiWLeuFWvh4wRAAQkR4wkW+ZG8iJ1FQAAPiJ0w0TYdHjAxTF-mxAFcUBfHEAJYbCaJlG6YmFj-DBUAACokLA0SekoFGoDRLBcOOAVoew6qFAhb5Sb6e5kAAcmQT6ydM3FHspkCqZRGnaRRDnZX5tEyWoRm-ix-4cYBPHAVZYhgUJUBxnZOliU5kkEF2UAAJKcJAs6rgaOAUNIwCBY6FDINI0ggAoUV0oEOA9cAEwQFWX70T+zGsexnHcbx0EzeF80zotED1qthnrSZW0VbxIE1YJ-j1eJe27nNC1LW2BlFRdpWmeV5mVXxP7KnVDXUfZzVQa1PpkGqCrSGQ6rKEID6JWoZIo0enHRhEZLY3J5gGEAA)

adamziel replied:
@dmsnell I added support for `Pick` to this PR. Unfortunately I couldn't figure out a way to retain the `Context` as the second generic parameter. This is the simplest signature I came up with:

```ts
	function <
		R extends RecordOf< K, N >,
		Query extends EntityQuery< K, N, InferredContext >,
		K extends Kind = KindOf< R >,
		N extends Name = NameOf< R >,
		InferredContext extends Context = Query extends { context: infer C }
			? C
			: DefaultContextOf< R >
	>(
		state: State,
		kind: K,
		name: N,
		key: KeyOf< R >,
		query?: Query
	): PickQueryFields< RecordOf< K, N, InferredContext >, Query > | undefined {
```

Since TS type inference is all or nothing, passing a type for R requires the query type, too:

```ts
const comment = getEntityRecord<
	Comment< 'view' >,
	{ context: 'view'; _fields: [ 'id' ] }
>( {}, 'root', 'comment', 15, { context: 'view', _fields: [ 'id' ] } );
// comment is Pick< Comment< 'view' >, 'id' >
```

Perhaps this isn't a big issue, though. I can't think of a case where you wouldn't want to rely on inference. `getEntityRecord` only supports the registered entity types with known kind and name and you can add new ones by extending the `PerPackageEntities` interface.

dmsnell replied:
leaving it at `Partial` for now is probably a good place to be.

in your playground link you can see the same trouble I faced when not using type-narrowing literals. Hint: if you pass your objects as literals in the function invocation or type parameter it will type differently than if you pass in an object defined elsewhere.

```ts
// this works
// Return the Picked version of an object
pick(t, { _fields: ['a', 'b'] });
```

```ts
// this doesn't work
const q = { _fields: [ 'a', 'b' ] };

// Return the Picked version of an object
pick(t, q);
```

Because it can't assert that `q._fields` are keys of the object. We can add some `readonly` attributes to the `(keyof Source)[]` types and that actually let's us do this, but there's a caveat. We have to define `q` with `as const`, or at lest the `_fields` property `as const`.

https://user-images.githubusercontent.com/5431237/156456188-6f24a6c6-3b86-4153-9cfc-54c93f43bf07.mov

There may still be a way to get this, but I consider it a non-starter if we have to have people declare those fields `as const` manually and in every invocation.

dmsnell replied:
The core problem I think is getting a list of strings into a type union of those string values without needing `as const`. This illustrates the problem as `Fields` and `Fields2` both resolve to `string` instead of as `'a' | 'b' | 'id' | 'name'`

```ts
const l = ['a', 'b', 'id', 'name'];

type Names<T extends any[]> = T[number];
type Names2<T extends any[]> = T extends (infer V)[] ? V : never;

type Fields = Names<typeof l>;
type Fields2 = Names2<typeof l>;
```

adamziel replied:
Ah great point @dmsnell! It seems like in the current TypeScript version all the type information is gone right after the variable declaration, so this isn't something we can fix. The best we can do here is return `Pick<>` when the keys are known, and `Partial<>` when they aren't. For now, let's abandon `Pick` and go with Partial as you suggested – `Pick` is a new feature that can be added later on.

dmsnell commented on packages/core-data/src/selectors-ts.ts:
```diff
@@ -0,0 +1,140 @@
+				? Partial< RecordOf< K, N, C > >
+				: RecordOf< K, N, C > )
+		| null
+		| undefined {
```
is `null` vs. `undefined` important? would it be better if we only returned one of those values?

adamziel replied:
There are some [empty test cases](https://github.com/WordPress/gutenberg/blob/8fbf70c6b6aa965af4eae4bb052b7a86acbdc1b9/packages/core-data/src/test/selectors.js#L28) indicating the `null` is a part of the de-facto public API signature of this function. If that's the case, removing it would be a BC breach.

That being said, I'd love if it just returned only, say, `null`. Let's proceed with `null | undefined` here to avoid getting blocked, and take this discussion to another PR.

adamziel committed:
Revert to using only Partial<> and not Pick<>

adamziel committed:
Remove dev artifacts and ignore ts error related to importing @wordpress/data

adamziel committed:
Remove unused import

adamziel committed:
Lint

adamziel committed:
Remove commented out test examples from selectors-ts.ts

adamziel committed:
Move getEntityRecord type definition to types/selectors.ts, use via @typedef for now

adamziel committed:
Add docstrings to EntityQuery and EntityFromConfig

adamziel committed:
Remove the unnecessary trailing comma

adamziel committed:
Restore the js version of get-normalized-comma-separable.js

adamziel committed:
Add type for getEditedEntityRecord

adamziel committed:
Remove Query from getEditedEntityRecord

adamziel committed:
Add getRawEntityRecord type

adamziel committed:
Wrap @type getEditedEntityRecord in braces

adamziel commented:
I moved the type definitions to types.js and plugged them in using `@typeDef`. I am not sure why the static analysis fails with `TypeError: Cannot read property '0' of undefined` – any ideas how can we fix it?

Edit: That error shows up after updating the comments from:

```js
/**
 * @param {Object}  state State tree
 * @param {string}  kind  Entity kind.
 * @param {string}  name  Entity name.
 * @param {number}  key   Record's key
 * @param {?Object} query Optional query.
 *
 * @return {Object?} Record.
 */
export const getEntityRecord = createSelector(
	( state, kind, name, key, query ) => {
```

to 

```js
/**
 * @type  {getEntityRecord}
 * 
 * @param  State tree
 * @param  kind  Entity kind.
 * @param  name  Entity name.
 * @param  key   Record's key
 * @param  query Optional query.
 *
 * @return Record.
 */
export const getEntityRecord = createSelector(
	( state, kind, name, key, query ) => {
```

I wonder how we could get around it

adamziel committed:
Reduce the public API surface area of the types to just the selectors, PerPackageEntities, and EntityRecord

adamziel committed:
Restore the trunk version of selectors.js

adamziel committed:
Make entity records aware of their default contexts

adamziel committed:
Use the same DefaultContextOf type for both by-record lookups and kind-name lookups

adamziel committed:
Add types for all core-data selectors

adamziel commented:
@dmsnell I removed the `@typeDef` annotations and added all the missing selectors signatures to `types/selectors.ts`. Combined with your work on TS-ifying core-data, it should be "just work" once `selectors.js` become `selectors.ts`.

I also added default context to Entity Record declarations, like this:

```ts
export type Comment<
	C extends Context = DefaultContextOf< 'root', 'comment' >
> = OmitNevers< _BaseEntityRecords.Comment< C > >;
```

adamziel committed:
Turn EntityQuery into a private type

adamziel committed:
Adjust import paths

adamziel committed:
Only return a partial object when _fields were passed

adamziel committed:
Use EntityQuery type in getEntityRecord type signature

adamziel committed:
Use trunk's version of selectors.js

adamziel committed:
Use a language consistent with the latest version of core-data

adamziel committed:
Remove the EntityDeclaration type

adamziel committed:
Remove the unused EntityInterface from entity-types/entities.ts

adamziel commented:
@dmsnell I rebased this on top of latest trunk, standardized the language a bit, and removed a few unused types.

adamziel committed:
Restore missing export keywords

adamziel commented:
This huge PR was entirely merged in small pieces listed in the description above. `@wordpress/core-data` selectors are now typed!

To make use of these type signatures useful outside of `core-data`, we'll also need to have a type signature for `useSelect`. Let's track this work separately – we have two PRs striving to do just that:

* https://github.com/WordPress/gutenberg/pull/39081
* https://github.com/WordPress/gutenberg/pull/41596
